runner/src/runner.rs

```diff
@@ -381,18 +360,69 @@ fn run_binary_with_retry(runner: &mut impl RunnerExt, stdout_mode: impl Into<Std
             // The `Transient` error type means the closure will be retried
             .map_err(BackoffError::Transient)
     }) {
-        Ok(child) => Ok(child),
+        Ok(child) => {
+            log::info!("Vault started, with pid {}", child.id());
+            Ok(child)
+        }
         Err(BackoffError::Permanent(err)) => Err(err.into()),
         Err(BackoffError::Transient(err)) => Err(err.into()),
     }
 }
 
+pub async fn run(mut runner: Box<dyn RunnerExt + Send>, mut shutdown_signals: SignalsInfo) -> Result<(), Error> {
+    tokio::select! {
+        _ = shutdown_signals.next() => {
+            log::info!("Terminating vault...");
+            runner.terminate_proc_and_wait()?;
+        },
+        _ = runner.as_mut().auto_update() => {
+            log::error!("Auto-updater unexpectedly terminated.");
+            return Err(Error::AutoUpdaterTerminated);
+        }
+    };
+    Ok(())
+}
+
+async fn auto_update(runner: &mut impl RunnerExt) -> Result<(), Error> {
```

Comment:
Personally I'm not fond of these standalone functions since their functionality is strictly tied to the `RunnerExt` type anyway.